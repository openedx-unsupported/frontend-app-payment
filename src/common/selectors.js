import { createSelector } from 'reselect';

export const userAccountSelector = state => state.userAccount;

export const emailSelector = createSelector(
  userAccountSelector,
  userAccount => userAccount.email,
);

export const configurationSelector = state => state.configuration;
