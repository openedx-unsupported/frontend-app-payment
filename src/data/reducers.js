import { combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';

import {
  reducer as paymentReducer,
  storeName as paymentStoreName,
} from '../payment';
import {
  reducer as feedbackReducer,
  storeName as feedbackStoreName,
} from '../feedback';

const createRootReducer = () => combineReducers({
  [paymentStoreName]: paymentReducer,
  [feedbackStoreName]: feedbackReducer,
  form: formReducer,
});

export default createRootReducer;
