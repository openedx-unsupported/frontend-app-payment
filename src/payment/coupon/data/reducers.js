import { ADD_COUPON, REMOVE_COUPON, UPDATE_COUPON_DRAFT } from './actions';

const defaultState = {
  benefit: null,
  code: null,
  voucherId: null,
  error: null,
  loaded: false,
  loading: false,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case ADD_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case ADD_COUPON.SUCCESS:
        return {
          ...state,
          loading: false,
          loaded: true,
          error: null,
          benefit: action.payload.benefit,
          code: action.payload.code,
          voucherId: action.payload.voucherId,
        };
      case ADD_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: false,
          error: action.payload.error,
          voucherId: null,
        };
      case REMOVE_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case REMOVE_COUPON.SUCCESS:
        return defaultState;
      case REMOVE_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: true, // There's still a loaded coupon.
          error: action.payload.error,
        };
      case UPDATE_COUPON_DRAFT:
        return {
          ...state,
          code: action.payload.code,
        };
      default:
    }
  }

  return state;
};

export default reducer;
