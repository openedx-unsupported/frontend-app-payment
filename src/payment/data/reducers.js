import { combineReducers } from 'redux';

import {
  BASKET_DATA_RECEIVED,
  BASKET_PROCESSING,
  CAPTURE_KEY_DATA_RECEIVED,
  CAPTURE_KEY_PROCESSING,
  MICROFORM_STATUS,
  fetchBasket,
  submitPayment,
  fetchCaptureKey,
} from './actions';

import { DEFAULT_STATUS } from '../checkout/payment-form/flex-microform/constants';

const basketInitialState = {
  loading: true,
  loaded: false,
  submitting: false,
  redirect: false,
  isBasketProcessing: false,
  products: [],
};

const basket = (state = basketInitialState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case fetchBasket.TRIGGER: return { ...state, loading: true };
      case fetchBasket.FULFILL: return {
        ...state,
        loading: false,
        loaded: true,
      };

      case BASKET_DATA_RECEIVED: return { ...state, ...action.payload };

      case BASKET_PROCESSING: return {
        ...state,
        isBasketProcessing: action.payload,
      };

      case submitPayment.TRIGGER: return {
        ...state,
        paymentMethod: action.payload.method,
      };
      case submitPayment.REQUEST: return {
        ...state,
        submitting: true,
      };
      case submitPayment.SUCCESS: return {
        ...state,
        redirect: true,
      };
      case submitPayment.FULFILL: return {
        ...state,
        submitting: false,
        paymentMethod: undefined,
      };

      default:
    }
  }
  return state;
};

const captureContextInitialState = {
  isCaptureKeyProcessing: false,
  microformStatus: DEFAULT_STATUS,
  captureKeyId: '',
};

const captureKey = (state = captureContextInitialState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case fetchCaptureKey.TRIGGER: return state;
      case fetchCaptureKey.FULFILL: return state;

      case CAPTURE_KEY_DATA_RECEIVED: return { ...state, ...action.payload };

      case CAPTURE_KEY_PROCESSING: return {
        ...state,
        isCaptureKeyProcessing: action.payload,
      };

      case MICROFORM_STATUS: return {
        ...state,
        microformStatus: action.payload,
      };

      default:
    }
  }
  return state;
};

const reducer = combineReducers({
  basket,
  captureKey,
});

export default reducer;
