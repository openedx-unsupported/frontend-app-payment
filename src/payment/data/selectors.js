import { createSelector } from 'reselect';
import {
  localeSelector,
  getCountryList,
} from '@edx/frontend-i18n';

export const storeName = 'payment';
export const paymentSelector = (state) => {
  const paymentState = state[storeName].basket;
  const loaded = !paymentState.loading && !paymentState.loadingError;
  return {
    ...paymentState,
    loaded,
    isEmpty: loaded && paymentState.products && paymentState.products.length > 0,
  };
};

export const countryOptionsSelector = createSelector(
  localeSelector,
  locale => ({
    countryOptions: getCountryList(locale).map(({ code, name }) => ({ value: code, label: name })),
  }),
);

export const basketSelector = state => ({ ...state[storeName].basket });

export const productsSelector = state => ({ ...state[storeName].basket });
