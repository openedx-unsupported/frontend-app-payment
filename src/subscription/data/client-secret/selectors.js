import { createSelector } from 'reselect';
import { storeName } from '../constants';

export const clientSecretSelector = state => (state[storeName] ? state[storeName].clientSecret : null);

export const getClientSecretSelector = createSelector(
  clientSecretSelector,
  (clientSecret, isClientSecretProcessing) => ({
    clientSecretId: clientSecret && clientSecret.capture_context ? clientSecret.capture_context.key_id : null,
    isClientSecretProcessing,
  }),
);
