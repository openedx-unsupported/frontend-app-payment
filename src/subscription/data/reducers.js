import { combineReducers } from 'redux';

import subscription from './details/reducer';
import clientSecret from './client-secret/reducer';

export const reducer = combineReducers({
  details: subscription,
  clientSecret,
});

export default reducer;
