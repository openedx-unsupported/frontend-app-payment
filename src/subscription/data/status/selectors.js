// import { createSelector } from 'reselect';
import { storeName } from '../constants';

// eslint-disable-next-line import/prefer-default-export
export const subscriptionStatusSelector = state => state[storeName].status;

// export const status = createSelector(
//   subscriptionStatusSelector,
//   status => status,
// );
