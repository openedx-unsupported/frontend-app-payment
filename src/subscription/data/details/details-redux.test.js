// import Cookies from 'universal-cookie';
import { createStore, combineReducers } from 'redux';

import { reducer } from '../reducers';
import {
  subscriptionDetailsReceived,
  subscriptionDetailsProcessing,
  submitPayment,
  fetchSubscriptionDetails,
} from './actions';
import { subscriptionSelector } from './selectors';

jest.mock('universal-cookie', () => {
  class MockCookies {
    static result = {};

    get() {
      return MockCookies.result;
    }
  }
  return MockCookies;
});

describe('subscription details redux tests', () => {
  let store;

  beforeEach(() => {
    store = createStore(combineReducers({
      subscription: reducer,
    }));
  });

  describe('selectors', () => {
    // describe('details', () => {
    it('default state', () => {
      const result = subscriptionSelector(store.getState());
      expect(result).toEqual({
        loading: true,
        loaded: false,
        submitting: false,
        redirect: false,
        enableStripePaymentProcessor: true,
        isSubscriptionDetailsProcessing: false,
        products: [],
        paymentMethod: 'stripe',
      });
    });
  });

  describe('details reducer', () => {
    it('should load the initialState when mount', () => {
      // Its base state
      expect(store.getState().subscription.details.loading).toBe(true);

      // If an action isn't appropriate for it
      store.dispatch({ type: 'SOMETHING_ELSE' });
      expect(store.getState().subscription.details.loading).toBe(true);
    });

    it('SUBSCRIPTION_DETAILS_RECEIVED action', () => {
      store.dispatch(subscriptionDetailsReceived({ foo: 'bar' }));
      expect(store.getState().subscription.details.foo).toBe('bar');
      expect(store.getState().subscription.details.loading).toBe(false);
      expect(store.getState().subscription.details.loaded).toBe(true);
    });

    describe('SUBSCRIPTION_DETAILS_PROCESSING action', () => {
      it('SUBSCRIPTION_DETAILS_PROCESSING true action', () => {
        store.dispatch(subscriptionDetailsProcessing(true));
        expect(store.getState().subscription.details.isSubscriptionDetailsProcessing).toBe(true);
      });

      it('BASKET_PROCESSING false action', () => {
        store.dispatch(subscriptionDetailsProcessing(false));
        expect(store.getState().subscription.details.isSubscriptionDetailsProcessing).toBe(false);
      });
    });

    describe('submitPayment actions', () => {
      const paymentProcessors = [
        'stripe',
      ];

      test.each(paymentProcessors)('submitPayment.TRIGGER action', (processor) => {
        store.dispatch(submitPayment({ method: processor }));
        expect(store.getState().subscription.details.paymentMethod).toBe(processor);
      });

      it('submitPayment.REQUEST action', () => {
        store.dispatch(submitPayment({ method: 'PayPal' }));
        expect(store.getState().subscription.details.paymentMethod).toBe('PayPal');
      });

      it('submitPayment.REQUEST action', () => {
        store.dispatch(submitPayment.request());
        expect(store.getState().subscription.details.submitting).toBe(true);
      });

      it('submitPayment.SUCCESS action', () => {
        store.dispatch(submitPayment.success());
        expect(store.getState().subscription.details.redirect).toBe(true);
      });

      it('submitPayment.FULFILL action', () => {
        store.dispatch(submitPayment.fulfill());
        expect(store.getState().subscription.details.submitting).toBe(false);
        expect(store.getState().subscription.details.paymentMethod).toBeUndefined();
      });
    });

    describe('fetchBasket actions', () => {
      test.only('fetchBasket.TRIGGER action', () => {
        store.dispatch(fetchSubscriptionDetails());
        expect(store.getState().subscription.details.loading).toBe(true);
      });

      it('fetchBasket.FULFILL action', () => {
        store.dispatch(fetchSubscriptionDetails.fulfill());
        expect(store.getState().subscription.details.loading).toBe(false);
        expect(store.getState().subscription.details.loaded).toBe(true);
      });
    });
  });
});
