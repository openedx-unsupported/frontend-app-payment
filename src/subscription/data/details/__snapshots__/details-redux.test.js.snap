// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscription details redux tests details reducer should return the default state when appropriate 1`] = `
Object {
  "certificateType": "verified",
  "currency": "USD",
  "isSubscriptionDetailsProcessing": false,
  "loaded": false,
  "loading": true,
  "organization": "University of California, Berkeley",
  "paymentMethod": "stripe",
  "price": 49,
  "products": Array [],
  "programTitle": "Blockchain Fundamentals",
  "redirect": false,
  "submitting": false,
}
`;

exports[`subscription details redux tests details reducer should return the default state when appropriate 2`] = `
Object {
  "certificateType": "verified",
  "currency": "USD",
  "isSubscriptionDetailsProcessing": false,
  "loaded": false,
  "loading": true,
  "organization": "University of California, Berkeley",
  "paymentMethod": "stripe",
  "price": 49,
  "products": Array [],
  "programTitle": "Blockchain Fundamentals",
  "redirect": false,
  "submitting": false,
}
`;

exports[`subscription details redux tests details reducer should return the default state when appropriate 3`] = `
Object {
  "clientSecret": Object {
    "clientSecretId": "",
    "isClientSecretProcessing": false,
  },
  "details": Object {
    "certificateType": "verified",
    "currency": "USD",
    "isSubscriptionDetailsProcessing": false,
    "loaded": false,
    "loading": true,
    "organization": "University of California, Berkeley",
    "paymentMethod": "stripe",
    "price": 49,
    "products": Array [],
    "programTitle": "Blockchain Fundamentals",
    "redirect": false,
    "submitting": false,
  },
}
`;
